packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -84,6 +84,49 @@ const Duration _kCursorBlinkHalfPeriod = Duration(milliseconds: 500);
 // is shown in an obscured text field.
 const int _kObscureShowLatestCharCursorTicks = 3;
 
+class _RenderCompositionCallback extends RenderProxyBox {
+  _RenderCompositionCallback(this.compositeCallback);
+
+  final void Function(Layer) compositeCallback;
+
+  bool get enabled => _enabled;
+  bool _enabled = false;
+  VoidCallback? _cancelCallback;
+  set enabled(bool newValue) {
+    if (!newValue) {
+      _cancelCallback?.call();
+      _cancelCallback = null;
+    }
+    _enabled = newValue;
+  }
+
+  @override
+  void paint(PaintingContext context, ui.Offset offset) {
+    if (enabled) {
+      _cancelCallback ??= context.addCompositionCallback(compositeCallback);
+    }
+    super.paint(context, offset);
+  }
+}
+
+class _CompositionCallback extends SingleChildRenderObjectWidget {
+  const _CompositionCallback({ required this.compositeCallback, this.enabled = false, super.child });
+  final void Function(Layer) compositeCallback;
+  final bool enabled;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    assert(!enabled);
+    return _RenderCompositionCallback(compositeCallback);
+  }
+  @override
+  void updateRenderObject(BuildContext context, _RenderCompositionCallback renderObject) {
+    super.updateRenderObject(context, renderObject);
+    assert(renderObject.compositeCallback == compositeCallback);
```

Comment:
How is this guaranteed?

Reply:
The class is private and doesn't have a global key, so it will never used by a different `EditableTextState`?